File: 0002-add-two-numbers/0002-add-two-numbers.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
const addTwoNumbers = (l1, l2) => {
  const sentinel = new ListNode(0);

  let tail = sentinel;
  let carry = 0;

  while (l1 !== null || l2 !== null || carry !== 0) {
    const x = (l1 !== null) ? l1.val : 0;
    const y = (l2 !== null) ? l2.val : 0;
    const sum = x + y + carry;

    const val = sum % 10;
    carry = Math.floor(sum / 10);

    tail.next = new ListNode(val);
    tail = tail.next;

    l1 = (l1 !== null) ? l1.next : null;
    l2 = (l2 !== null) ? l2.next : null;
  }

  return sentinel.next;
};

File: 0002-add-two-numbers/0002-add-two-numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        sentinel = ListNode()
        tail = sentinel
        current1 = l1
        current2 = l2
        carry = 0

        while current1 or current2 or carry:
            second_value = current1.val if current1 else 0
            first_value = current2.val if current2 else 0
            total_sum = first_value + second_value + carry

            new_node_value = total_sum % 10
            carry = total_sum // 10

            tail.next = ListNode(new_node_value)

            tail = tail.next
            current1 = current1.next if current1 else None
            current2 = current2.next if current2 else None

        return sentinel.next
